pkg/cache/multi_namespace_cache.go

```diff
@@ -37,37 +39,154 @@ type NewCacheFunc func(config *rest.Config, opts Options) (Cache, error)
+}
+
+// MultiNamespacedCacheWithOptionsBuilder builds a composite cache that delegates to per-namespace
+// caches built according to the passed MultiNamespacedOption options.
+//
+// If the set of options passed to MultiNamespacedCacheWithOptionsBuilder results in the NewCacheFunc
+// being set multiple times for the same namespace, the last such setting will be used.
+//
+// If a default namespaced cache is defined (e.g. with WithDefaultNamespacedCache), it will be used
+// as a catch-all for objects not in the explicit namespaces configured on the cache builder. The
+// default namespaced cache is automatically configured with extra field selectors to avoid duplicate
+// caching of objects between namespace-specific caches and this catch-all cache.
+//
+// If a cluster-scoped cache is defined (e.g. with WithClusterScopedCache), it will be used for
+// cluster-scoped objects. If it is undefined, the resulting cache will return an error during read
+// operations, reporting that a cluster-scoped cache is not defined.
+func MultiNamespacedCacheWithOptionsBuilder(opts ...MultiNamespacedOption) NewCacheFunc {
+	multiNamespaceOpts := MultiNamespacedOptions{
+		NewNamespacedCaches: map[string]NewCacheFunc{},
+	}
+	for _, opt := range opts {
+		opt(&multiNamespaceOpts)
+	}
+
 	return func(config *rest.Config, opts Options) (Cache, error) {
 		opts, err := defaultOpts(config, opts)
 		if err != nil {
 			return nil, err
 		}
 
-		caches := map[string]Cache{}
+		var clusterCache Cache
+		if multiNamespaceOpts.NewClusterScopedCache != nil {
+			clusterCache, err = multiNamespaceOpts.NewClusterScopedCache(config, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
 
-		// create a cache for cluster scoped resources
-		gCache, err := New(config, opts)
-		if err != nil {
-			return nil, fmt.Errorf("error creating global cache: %w", err)
+		nsToCache := map[string]Cache{}
+		if multiNamespaceOpts.NewDefaultNamespacedCache != nil {
+			defaultNamespacedOpts := setDefaultNamespacedCacheOpts(opts, multiNamespaceOpts.NewNamespacedCaches)
+			defaultNamespacedCache, err := multiNamespaceOpts.NewDefaultNamespacedCache(config, defaultNamespacedOpts)
+			if err != nil {
+				return nil, err
+			}
+			nsToCache[corev1.NamespaceAll] = defaultNamespacedCache
 		}
 
-		for _, ns := range namespaces {
+		for ns, newCacheFunc := range multiNamespaceOpts.NewNamespacedCaches {
 			opts.Namespace = ns
-			c, err := New(config, opts)
+			nsToCache[ns], err = newCacheFunc(config, opts)
 			if err != nil {
 				return nil, err
 			}
-			caches[ns] = c
 		}
-		return &multiNamespaceCache{namespaceToCache: caches, Scheme: opts.Scheme, RESTMapper: opts.Mapper, clusterCache: gCache}, nil
+
+		return &multiNamespaceCache{
+			namespaceToCache: nsToCache,
+			clusterCache:     clusterCache,
+			RESTMapper:       opts.Mapper,
+			Scheme:           opts.Scheme,
+		}, nil
+	}
+}
+
+func setDefaultNamespacedCacheOpts(opts Options, newObjectCaches map[string]NewCacheFunc) Options {
```

Comment:
So this will end up setting up a cluster-scoped cache with a field-selector that excludes all namespaces for which we have a cache?

If yes, the problem with that is that it will fail if you don't have cluster-wide rbac - Wouldn't it be more useful to create a namespace-scoped cache with the configured selector whenever we get a request for a namespace that doesn't have a cache?

Reply:
> So this will end up setting up a cluster-scoped cache with a field-selector that excludes all namespaces for which we have a cache?

It depends on the NewCacheFunc defined by `WithDefaultNamespacedCache`. But yeah, I'd generally imagine that would be a cluster-wide cache.

> Wouldn't it be more useful to create a namespace-scoped cache with the configured selector whenever we get a request for a namespace that doesn't have a cache?

Two thoughts:
1. If a caller uses `client.List()` without `client.InNamespace` they're essentially asking for all $things in the cluster (modulo the selectors used to configure the various subcaches). It isn't immediately obvious to me how we'd handle this situation with individual per-namespace caches for catch-all namespaces.
2. There's a performance/permissions tradeoff here. If you assume cluster-wide list/watch RBAC, you only need one informer. If you don't/can't assume cluster-wide list/watch RBAC, you'd end up creating an informer and open a watch connection for each catch-all namespace you end up querying. In a cluster with lots of namespaces, that could end up putting quite a load on the apiserver.

Thoughts? Perhaps we could make this configurable?

Reply:
> If a caller uses client.List() without client.InNamespace they're essentially asking for all $things in the cluster (modulo the selectors used to configure the various subcaches). It isn't immediately obvious to me how we'd handle this situation with individual per-namespace caches for catch-all namespaces.

I think the expectation there would be that you get whatever is in the cache, which is a subset of what is actually in the cluster

> There's a performance/permissions tradeoff here. If you assume cluster-wide list/watch RBAC, you only need one informer. If you don't/can't assume cluster-wide list/watch RBAC, you'd end up creating an informer and open a watch connection for each catch-all namespace you end up querying. In a cluster with lots of namespaces, that could end up putting quite a load on the apiserver.

But what would be the reason for using this cache in the first place if you have cluster-wide rbac? Just getting a selector for one namespace only? Or is the actual, hidden use-case behind this "namespaced cache for object type X, global cache for everything else"?

I feel like the more we start supporting such rather exotic use-cases, the more ppl will just come up with even more of them. What if the next person wants to have different cache configurations based on object type or a combination of object type and namespace? Wouldn't that be the point where it would be easier to have some kind of `CacheRouter` that allows to freely configure something that inspects requests and then forwards them to the appropriate, configurable cache?

Sorry for being a bit negative here, it is just that this is IMHO both complicated enough to not be obvious to reason about and not solving the problem of "People might want an arbitrary cache depending on any possible combination of object type and selectors".

Reply:
> But what would be the reason for using this cache in the first place if you have cluster-wide rbac?

The specific motivation I have for this cache is memory optimization. I want to cache all configmaps in my controllers namespace, but in every other namespace, I only want to cache configmaps that match a certain label selector.